pkg/cmd/dcos_cluster_list.go

```diff
@@ -0,0 +1,122 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dcos/dcos-cli/pkg/cli"
+	"github.com/spf13/cobra"
+)
+
+// newCmdClusterList lists the clusters.
+func newCmdClusterList(ctx *cli.Context) *cobra.Command {
+	var onlyAttached bool
+	var jsonOutput bool
+	cmd := &cobra.Command{
+		Use:  "list",
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clusters, err := ctx.Clusters()
+			if err != nil {
+				return err
+			}
+
+			currentCluster, err := ctx.Cluster()
+			if err != nil {
+				return err
+			}
+			currentClusterID := currentCluster.ID()
+
+			// clustersData is composed of clusters containing its name, id, status, version, and url.
+			infos := []clusterInfo{}
+			var mu sync.Mutex
+
+			var wg sync.WaitGroup
+			for _, cluster := range clusters {
+				// Add should execute before the statement creating the goroutine.
+				wg.Add(1)
+				go func(cluster *cli.Cluster) {
+					defer wg.Done()
+
+					var info clusterInfo
+					info.Name = cluster.Name()
+					info.ID = cluster.ID()
+					if info.ID == currentClusterID {
+						info.Attached = true
+						if !jsonOutput {
+							info.Name += "*"
+						}
+					} else if onlyAttached {
+						return
+					}
+					info.URL = cluster.URL()
+					info.Status = "AVAILABLE"
+					info.Version, err = getVersion(ctx, cluster)
+					if err != nil {
+						info.Status = "UNAVAILABLE"
+					}
+
+					mu.Lock()
+					infos = append(infos, info)
+					mu.Unlock()
+				}(cluster)
+			}
+			wg.Wait()
+
+			if jsonOutput {
+				// Re-marshal it into json with indents added in for pretty printing.
+				out, err := json.MarshalIndent(infos, "", "\t")
+				if err != nil {
+					return err
+				}
+				fmt.Fprintln(ctx.Out, string(out))
+			} else {
+				table := cli.NewTable(ctx.Out, []string{"NAME", "ID", "STATUS", "VERSION", "URL"})
+				for i := range infos {
+					table.Append([]string{infos[i].Name, infos[i].ID, infos[i].Status, infos[i].Version, infos[i].URL})
+				}
+				table.Render()
+			}
+			return nil
+		},
+	}
+	cmd.Flags().BoolVar(&onlyAttached, "attached", false, "returns attached clusters only")
+	cmd.Flags().BoolVar(&jsonOutput, "json", false, "returns clusters in json format")
+	return cmd
+}
+
+// getVersion returns the version of the cluster as a string.
+func getVersion(ctx *cli.Context, cluster *cli.Cluster) (string, error) {
+	if cluster.URL() == "" {
+		return "UNKNOWN", errors.New("cluster URL undefined")
+	}
+
+	client := ctx.HTTPClient(cluster)
+	resp, err := client.Get("/dcos-metadata/dcos-version.json")
```

Comment:
This might hang for some time under some circumstances, could we hardcode a 5 seconds timeout here?

`client := ctx.HTTPClient(cluster, httpclient.Timeout(5 * time.Second))` on the line above.